src/Components/Components/src/NavigationException.cs

```diff
@@ -11,7 +11,7 @@ public class NavigationException : Exception
     /// <summary>
     /// Initializes a new <see cref="NavigationException"/> instance.
     /// </summary>
-    public NavigationException(string uri)
+    public NavigationException(string uri) : base($"Could not navigate to {uri}")
```

Comment:
I would propose we update the error message to include the reason why the exception is being seen by the developer in the first place. This exception is somewhat unique in that its purpose is to signal to the framework that a redirect should happen (rather than indicate an error occurred), so it might be best if the exception message reflected that.

Something like:

> "A navigation was initiated during static rendering. This exception is not an error, but instead signals to the framework that a redirect should occur. It should not be caught by application code."

However, I'm not sure that we should use that message for _all_ `NavigationException` instances, because there's nothing about the `NavigationException` type that says it must only be thrown for the sake of redirecting during static server rendering. The current XML docs for `NavigationException` suggest that it gets thrown when the navigation manager "is not able to navigate to a different URL", so it's not totally unreasonable to think that a custom `NavigationManager` implementation somewhere out there could have thrown the exception for a different reason. We _could_ change the XML docs and indicate that the exception should not be _thrown_ by application code either, but we'd need to have a discussion about how "breaking" we think that is.

I think the best option is to add an overload to the `NavigationException` constructor that accepts a `message`. Then we change each internal call site in the framework to pass an explicit message to the `NavigationException` constructor.

cc @SteveSandersonMS @javiercn @danroth27 in case you have other perspectives

Reply:
> I think the best option is to add an overload to the NavigationException constructor that accepts a message. Then we change each internal call site in the framework to pass an explicit message to the NavigationException constructor.

Sounds reasonable to me.

Reply:
Thanks @MackinnonBuck for the thorough answer.

Should I go ahead and work on this fix?

Reply:
@carlos-regis Yes, that would be great! And I don't think it will be necessary to include the URI in the message.

We currently repeat the following two lines in at least three different places:
https://github.com/dotnet/aspnetcore/blob/76a4e7507352de5cbabfd8d0130b69012442fe3f/src/Components/Endpoints/src/DependencyInjection/HttpNavigationManager.cs#L14-L15

(e.g., [here](https://github.com/dotnet/aspnetcore/blob/76a4e7507352de5cbabfd8d0130b69012442fe3f/src/Components/Server/src/Circuits/RemoteNavigationManager.cs#L90-L91) and [here](https://github.com/dotnet/aspnetcore/blob/76a4e7507352de5cbabfd8d0130b69012442fe3f/src/Components/Server/src/Circuits/RemoteNavigationManager.cs#L131-L132))

...So we'll need to update the exception message for each of those call sites.

Maybe we could introduce a [shared source file](https://github.com/dotnet/aspnetcore/tree/main/src/Components/Shared/src) containing a helper that converts the given URI string to an absolute URI and throws a `NavigationException` with that URI and the updated message. Something like...

```csharp
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

namespace Microsoft.AspNetCore.Components;

internal static class NavigationRedirectHelper
{
    private const string RedirectExceptionMessage =
        "A navigation was initiated during static rendering. " +
        "This exception is not an error, but instead signals to the framework that a redirect should occur. " +
        "It should not be caught by application code.";

    [DoesNotReturn]
    public static void ThrowNavigationExceptionForRedirect(NavigationManager navigationManager, string uri)
    {
         var absoluteUriString = ToAbsoluteUri(uri).AbsoluteUri; 
         throw new NavigationException(absoluteUriString, RedirectExceptionMessage); 
    }
}
```

...which we then use at each of the places where we currently throw a `NavigationException`.

(disclaimer: I haven't tried compiling this so the real code may be slightly different)

Reply:
@MackinnonBuck I am working on it and will get back to you soon.

Reply:
Hi @carlos-regis, just checking in - is this a change you are still interested in contributing? Please let us know either way so that the Blazor team knows whether a core team member should take this PR across the finish line. This is a change we think will be very valuable!